lib/puppet/agent.rb

```diff
@@ -1,6 +1,167 @@
 require 'puppet/application'
 require 'puppet/error'
 
+# A class providing pre/post fork callbacks modeled after those registered
+# by the `pthread_atfork` function.
+# The callbacks defined in this class are no-ops on all platforms except
+# Solaris where they ensure the forked process runs in a different contract
+# than the parent process. This is necessary in order for the child process
+# (an agent run) to be able to restart/stop the puppet agent service without
+# being killed (as a result of the termination of the parent process
+# contract during the service restart/stop).
+# The implementation uses Fiddle to make calls to functions defined in the
+# libcontract library.
+class AtFork
+  require 'fiddle'
+
+  raise LoadError, %q(The loaded fiddle version doesn't define 'Fiddle::Handle') unless defined?(Fiddle::Handle)
```

Comment:
A question: What should we do on Solaris when Fiddle is not available or too old?
Should we:
1. Fallback to no-op? Rationale: It's how it used to work and the chances are hight the code won't be restarting the service and thus won't be affected by the problem we are trying to solve.
2. Raise an exception? Rationale: It either works in all scenarios as it supposed to or not at all ...